### PR TITLE
fix: heroku sub-generator improvement.

### PR DESCRIPTION
**Pull request by zhyd1997**
Fix: #20541 
---

Please make sure the below checklist is followed for Pull Requests.

- [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [ ] Tests are added where necessary
- [ ] The JDL part is updated if necessary
- [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [ ] Documentation is added/updated where necessary
- [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

**Review by mshima — COMMENTED** [review]
Please use await/async instead of events.
Events will not block and by the time it is emitted, the generator may have continued.

prefer `this.spawnCommand` to execute commands.

File: generators/heroku/generator.mjs [hunk 1]
```diff
@@ -709,7 +709,48 @@ export default class HerokuGenerator extends BaseGenerator {
...
             this.logger.log(chalk.bold('\nConfiguring Heroku'));
             await execCmd(`heroku config:set ${configVars}--app ${this.herokuAppName}`);
-            await execCmd(`heroku buildpacks:add ${buildpack} --app ${this.herokuAppName}`);
+            const { stdout: data } = await execCmd(`heroku buildpacks:add ${buildpack} --app ${this.herokuAppName}`);
+            if (data) {
+              this.logger.info(data);
+              // remote:  !     The following add-ons were automatically provisioned: . These add-ons may incur additional cost,
+              // which is prorated to the second. Run `heroku addons` for more info.
+              if (data.includes('Run `heroku addons` for more info.')) {
+                this.spawnCommand('heroku addons');
```
Comment (mshima) [inline generators/heroku/generator.mjs:718]: await it?

File: generators/heroku/generator.mjs [hunk 2]
```diff
@@ -709,7 +709,48 @@ export default class HerokuGenerator extends BaseGenerator {
...
+              ];
+
+              this.log('');
+              const props = await this.prompt(prompts);
+              if (props.userDeployDecision === 'Yes') {
+                this.log.info(chalk.bold('Continued deploying...'));
+              } else {
+                this.log(this.logger);
+                this.log.info(chalk.bold('You aborted deployment!'));
+                this.abort = true;
```
Comment (mshima) [inline generators/heroku/generator.mjs:748]:
```suggestion
                throw new Error('relevant message');
```

File: generators/heroku/generator.mjs (hunk 1 above)
Comment (zhyd1997) [reply generators/heroku/generator.mjs:718]: Tested and `this.spawnCommand` throw errors (even with `await`), should use `await execCmd('heroku addons')` back.

Screenshot 2023-04-23 at 19 44 07

File: generators/heroku/generator.mjs (hunk 2 above)
Comment (zhyd1997) [reply generators/heroku/generator.mjs:748]: Thanks for your suggestions, but here `throw new Error` will not abort program, `return` is needed.

Screenshot 2023-04-23 at 19 33 38

**Comment by mraible** [comment]
@mshima Is this one OK to merge?

File: generators/heroku/generator.mjs [hunk 3]
```diff
@@ -709,7 +709,48 @@ export default class HerokuGenerator extends BaseGenerator {
...
+              this.log('');
+              const prompts = [
+                {
+                  type: 'list',
+                  name: 'userDeployDecision',
+                  message: 'Continue to deploy?',
+                  choices: [
+                    {
+                      value: 'Yes',
+                      name: 'Yes, I confirmed',
```
Comment (mraible) [inline generators/heroku/generator.mjs:730]:
```suggestion
                      name: 'Yes, I confirm',
```

File: generators/heroku/generator.mjs [hunk 4]
```diff
@@ -709,7 +709,48 @@ export default class HerokuGenerator extends BaseGenerator {
...
+                    },
+                    {
+                      value: 'No',
+                      name: 'No, break (Recommended)',
```
Comment (mraible) [inline generators/heroku/generator.mjs:734]:
```suggestion
                      name: 'No, abort (Recommended)',
```

File: generators/heroku/generator.mjs [hunk 5]
```diff
@@ -709,7 +709,48 @@ export default class HerokuGenerator extends BaseGenerator {
...
+                    },
+                  ],
+                  default: 0,
+                },
+              ];
+
+              this.log('');
+              const props = await this.prompt(prompts);
+              if (props.userDeployDecision === 'Yes') {
+                this.log.info(chalk.bold('Continued deploying...'));
```
Comment (mraible) [inline generators/heroku/generator.mjs:744]:
```suggestion
                this.log.info(chalk.bold('Continuing deployment...'));
```

File: generators/heroku/generator.mjs [hunk 6]
```diff
@@ -710,7 +710,48 @@ export default class HerokuGenerator extends BaseGenerator {
...
+                    {
+                      value: 'Yes',
+                      name: 'Yes, I confirm',
+                    },
+                    {
+                      value: 'No',
+                      name: 'No, abort (Recommended)',
+                    },
+                  ],
+                  default: 0,
```
Comment (atomfrede) [inline generators/heroku/generator.mjs:738]: If abort is the recommended option why using '0' aka yes as default option? Besides this it looks fine to me, but didn't do proper testing yet (as my local development setup is broken for whatever reason).

File: generators/heroku/generator.mjs (hunk 2 above)
Comment (mshima) [reply generators/heroku/generator.mjs:749]: Ok, it's inside a callback.

**Comment by mshima** [comment]
> @mshima Is this one OK to merge?

Feel free to merge whenever you finish reviewing.

**Review by DanielFran — APPROVED** [review]

**Comment by DanielFran** [comment]
Thanks @zhyd1997 for the contribution

**Pull request merged**